File: Arrays_and_Hashing/Top_K_Frequent_Elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    let map =  new Map()

    nums.forEach(num => {
      map.set(num, map.get(num) + 1 || 1);
    });

    let sortedArray = [];

    for (let [key, value] of map) {
      sortedArray.push([key, value]);
    }

    sortedArray.sort((a, b) => b[1] - a[1]);

    let result = [];
    
    for (i = 0; i < k; i++) {
      result.push(sortedArray[i][0]);
    }
    
return result;
};

File: Arrays_and_Hashing/TopKFrequentElements.go
func topKFrequent(nums []int, k int) []int {
    bucket := make([][]int, len(nums)+1)
    countMap := make(map[int]int)

    for _, val := range nums {
        countMap[val]++
    }

    for key, val := range countMap {
        bucket[val] = append(bucket[val], key)
    }

    result := []int{}
    for i := len(bucket) -1; i > 0; i-- {
        for _, val := range bucket[i] {
            result = append(result, val)
            if len(result) == k {
                return result
            }
        }
    }
    return result
}

// Keep track of each number and their count
// Keep the count as an array with the indices being the number of occurences.
// Iterate through this array starting from the back until we hit k values 
